33 m partition basic methods

alanarteagav opened the pull request:
This PR from branch 33-m-partition-basic-methods implements the following four methods described in #33 :

- [x]  `IsClique`
- [x]  `IsStable` 
- [x]  `AreFullyAdjacent` 
- [x]  `AreFullyNonAdjacent` 

## verifiers.go

This Go source file contains the implementation for the aforementioned methods. 

## intSet.go

This file contains a convenience implementation of integer set: `IntSet`. The aim of this implementation is both to refactor and make more verbose the `efficientAdjacencyList` auxiliar method in `staticGraph.go`, and to provide a reliable set implementation for future GATTO's issues.

## graph.go, staticGraph.go

These files contain both the declaration and implementation of a tentative new  method for the `graph` interface: `Neighbours(v int)`. The method receives a vertex _v_ and returns a integer set (`IntSet`) of the neighbors of _v_, whether the graph is defined for an adjacency matrix or an adjacency list.


## sliceUtils.go

DisjointIntSlices & WithinIntervalSlice convenience methods proposal.

Japodrilo commented:
Hi @alanarteagav:

The work within the PR looks great, I think the auxiliary functions are a great idea.

I have some observations/questions:
- For the methods `IsClique` and `IsStable`, you check whether the neighbourhood of each vertex contains every other vertex in the set.  Would it be possible/more efficient to remove the first item from `vertices`?   You have already checked that it is adjacent/non-adjacent to every other vertex, so right now you are double-checking every vertex in `vertices`.
- It would be nice to have tests for `IntSet`.
- I think that in the future we might need the neighbours of a vertex as a list.   Maybe we should rename the method you added in `graph.go` to `NeighboursSet`, what do you think?

Thank you!

Japodrilo reviewed (changes requested):
Hi @alanarteagav:

Sorry, I just commented on the PR but forgot to leave my review.   Just address the short suggestions I added in the comments, and this is ready to go!

alanarteagav commented:
Hi @Japodrilo:

I have just addressed the remaining suggestions:

- Added unit tests for `IntSet` in `intSet_test.go`.
- Improved performance in `IsClique` & `IsStable` methods.
- Renamed `Neighbours` to `NeighboursSet` method.

The current commit passes all test cases.

Japodrilo reviewed (commented):
Hi @alanarteagav:

I am very sorry for the delay.   Great job!  This looks very good.

I can tell that the methods `IsClique`, `IsStable`, `AreFullyAdjacent`, and `AreFullyNonAdjacent` are efficient and work with both matrix and list representations. I am just wondering, if we already know that we are working with the matrix representation (e.g., by checking whether the matrix of the graph is not nil), would not it be faster just to iterate the input slices with a for loop, and just check whether every obtained entry (i,j) equals 1 in the adjacency matrix?   I guess that, even though the input does not correspond to a consecutive interval of integers, you should be able to check "just half" of the pairs.

Does this makes sense?

Japodrilo reviewed (approved):
Hi @alanarteagav:

The improvements look great, thank you!   Just a final detail.   In the tests for `IsClique`, `IsStable`, `AreFullyAdjacent`, and `AreFullyNonAdjacent`, there is one case which is not covered, this is, when the values of the subset are outside the range of the vertex set.   Just add such a verification, and feel free to merge this branch without further revision.

Japodrilo reviewed (approved):
Hi @alanarteagav:

Everything looks good.  Thank you!   This is now ready for merging.

Japodrilo reviewed (approved):
Hi @alanarteagav,

Sorry for the delay -_-'   Everything looks good.   There was a merge from `main` into this branch, which was causing duplicate commits when rebasing.   So, I removed that merge, and rebased with respect to `main`.   Unfortunately, this could not be done without force pushing my local branch.   So, you must delete your local branch if you want to pull the final changes.   Otherwise, you are ready to merge this branch, and just delete it.

In the future, please use `pull --rebase` instead of merging `main` into a feature branch.   Most of the times, you can rebase instead of merging (a notable exception is when we are merging a feature branch into `main` after rebasing).
